astra/src/main/java/com/slack/astra/clusterManager/CacheNodeAssignmentService.java

```diff
@@ -0,0 +1,568 @@
+                          cacheNodeAssignment,
+                          Metadata.CacheNodeAssignment.CacheNodeAssignmentState.EVICT);
+
+                  addCallback(
+                      future,
+                      successCountingCallback(successCounter),
+                      MoreExecutors.directExecutor());
+                  return future;
+                })
+            .collect(Collectors.toUnmodifiableList());
+
+    ListenableFuture<?> futureList = Futures.successfulAsList(replicaEvictions);
+    try {
+      futureList.get(futuresListTimeoutSecs, TimeUnit.SECONDS);
+    } catch (Exception e) {
+      futureList.cancel(true);
+    }
+
+    int successfulEvictions = successCounter.get();
+    int failedEvictions = replicaEvictions.size() - successfulEvictions;
+
+    assignmentCreateSucceeded.increment(successfulEvictions);
+    assignmentCreateFailed.increment(failedEvictions);
+
+    long evictionDuration = evictionTimer.stop(evictAssignmentTimer);
+    LOG.info(
+        "Completed assignment evictions - successfully marked {} assignments for eviction, failed to mark {} assignments for eviction in {} ms",
+        successfulEvictions,
+        failedEvictions,
+        nanosToMillis(evictionDuration));
+
+    return successfulEvictions;
+  }
+
+  /**
+   * Calculates and persists the HPA (Horizontal Pod Autoscaler) metric based on new assignments.
+   * This method calculates the demand factor from new assignments and persists it in the HPA metric
+   * metadata store.
+   *
+   * @param hpaMetricMetadataStore the store to persist HPA metrics
+   * @param replicaSet the replica set for which to calculate the HPA metric
+   * @param newAssignments the new assignments of cache nodes to snapshots
+   * @param cacheNodes the list of cache nodes
+   */
+  private static void calculateAndPersistHpaMetric(
+      HpaMetricMetadataStore hpaMetricMetadataStore,
+      String replicaSet,
+      Map<String, CacheNodeBin> newAssignments,
+      List<CacheNodeMetadata> cacheNodes) {
+    double demandFactor =
+        calculateHpaValueFromNewAssignments(
+            newAssignments, cacheNodes.stream().mapToLong(node -> node.nodeCapacityBytes).sum());
+    persistCacheConfig(hpaMetricMetadataStore, replicaSet, demandFactor);
+  }
+
+  /**
+   * Persists cache node assignments in the cache node assignment store.
+   *
+   * @param cacheNodeAssignmentStore the store to persist cache node assignments
+   * @param newAssignments a map of new assignments keyed by cache node ID
+   * @param replicaSet the replica set for which to persist assignments
+   * @return the number of successful assignments
+   */
+  private static int createAssignments(
+      CacheNodeAssignmentStore cacheNodeAssignmentStore,
+      Map<String, CacheNodeBin> newAssignments,
+      String replicaSet,
+      Map<String, ReplicaMetadata> replicasBySnapshotId) {
+    int numAssigned = 0;
+    for (Map.Entry<String, CacheNodeBin> entry : newAssignments.entrySet()) {
```

Comment:
Something you should keep in mind is HashMap `entrySet` isn't guaranteed to have any order. If order is important for any of the assignment logic you should look into using a different map implementation, or a list.